src/client.ts

```diff
@@ -282,62 +282,73 @@ export function createClient(options: ClientOptions): Client {
   > {
     locks++;
 
-    const socket = await (connecting ??
-      (connecting = new Promise<WebSocket>((resolve, reject) =>
-        (async () => {
-          if (retrying) {
-            await retryWait(retries);
-            retries++;
-          }
-
-          emitter.emit('connecting');
-          const socket = new WebSocketImpl(url, GRAPHQL_TRANSPORT_WS_PROTOCOL);
-
-          socket.onclose = (event) => {
-            connecting = undefined;
-            emitter.emit('closed', event);
-            reject(event);
-          };
-
-          socket.onopen = async () => {
-            try {
-              socket.send(
-                stringifyMessage<MessageType.ConnectionInit>({
-                  type: MessageType.ConnectionInit,
-                  payload:
-                    typeof connectionParams === 'function'
-                      ? await connectionParams()
-                      : connectionParams,
-                }),
-              );
-            } catch (err) {
-              socket.close(
-                4400,
-                err instanceof Error ? err.message : new Error(err).message,
-              );
+    const [socket, waitForCompleteOrThrowOnClose] = await (connecting ??
+      (connecting = new Promise<[WebSocket, Promise<void>]>(
+        (connected, denied) =>
+          (async () => {
+            if (retrying) {
+              await retryWait(retries);
+              retries++;
             }
-          };
-
-          socket.onmessage = ({ data }) => {
-            socket.onmessage = null; // interested only in the first message
-            try {
-              const message = parseMessage(data);
-              if (message.type !== MessageType.ConnectionAck) {
-                throw new Error(
-                  `First message cannot be of type ${message.type}`,
+
+            emitter.emit('connecting');
+            const socket = new WebSocketImpl(
+              url,
+              GRAPHQL_TRANSPORT_WS_PROTOCOL,
+            );
+
+            socket.onclose = (event) => {
+              connecting = undefined;
+              emitter.emit('closed', event);
+              denied(event);
+            };
+
+            socket.onopen = async () => {
+              try {
+                socket.send(
+                  stringifyMessage<MessageType.ConnectionInit>({
+                    type: MessageType.ConnectionInit,
+                    payload:
+                      typeof connectionParams === 'function'
+                        ? await connectionParams()
+                        : connectionParams,
+                  }),
+                );
+              } catch (err) {
+                socket.close(
+                  4400,
```

Comment:
some reference where I can find out the meaning of the codes sent on socket close?

Reply:
The 4XXX close event codes are reserved for applications. For convenience, I use HTTP status codes prefixed with a 4.

4400 can be seen as a HTTP `400 Bad Request`.